src/primitives/transaction.h

```diff
@@ -610,7 +625,27 @@ friend class CTransaction;
 
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        READWRITE(withdrawalEpochLength);
+
+        // The first serialization field is a 4 bytes element containing the SC Creation version in the first byte
+        // and the withdrawalEpochLenght in the remaining 3 bytes. The first implementation didn't have the version
+        // field, but only the withdrawalEpochLenght, so the version was implicitly set to 0 anyway.
+        if (ser_action.ForRead())
+        {
+            int withdrawalEpochLengthAndVersion;
+            READWRITE(withdrawalEpochLengthAndVersion);
+            
+            // Get the most significant byte
+            version = withdrawalEpochLengthAndVersion >> 24;
+
+            // Get the least significant 3 bytes
+            withdrawalEpochLength = withdrawalEpochLengthAndVersion & 0x00FFFFFF;
+        }
+        else
+        {
+            int withdrawalEpochLengthAndVersion = (version << 24) | withdrawalEpochLength;
```

Comment:
I think that we can skip the check for the `version` field due to its `uint8_t` datatype (so it will never be greater than 0xFF). It might instead be useful for the `withdrawalEpochLength` since it could overwrite the `version` during the OR bitwise operation, but I'm not sure if it's possible also in practice since the `withdrawalEpochLength` is limited by the chain parameters (see `nScMaxWithdrawalEpochLength` in `chainparams.cpp`).

Do you have any specific scenario in mind that we should take care of?

Reply:
Agree about version. For what regards withdrawalEpochLength the mask (or assert) would be useful to detect / prevent buggy clients generating messy txs (eg. when tx is not yet in a block / mempool then not validated yet by consensus rules).